R/counts.R

```diff
@@ -180,6 +179,151 @@ get_single_cell_experiment <- function(
+#'   [get_metadata()] function.
+#' @param assays A character vector whose elements must be either "counts"
+#'   and/or "cpm", representing the corresponding assay(s) you want to request.
+#'   By default only the count assay is downloaded. If you are interested in
+#'   comparing a limited amount of genes, the "cpm" assay is more appropriate.
+#' @param repository A character vector of length one. If provided, it should be
+#'   an HTTP URL pointing to the location where the summarized experiment data is stored.
+#' @param cache_directory An optional character vector of length one. If
+#'   provided, it should indicate a local file path where any remotely accessed
+#'   files should be copied.
+#' @param features An optional character vector of features (ie genes) to return
+#'   the counts for. By default counts for all features will be returned.
+#' @returns A SummarizedExperiment object, with one assay for each value in the
+#'   assays argument
+#' @examples
+#' \dontrun{
+#' meta <- get_metadata() |> 
+#'         filter(sample_ == '068502277538ef5559154b543167fefa' | 
+#'                sample_ == '07605571f51519f71da03704f056fa43')
+#' sme <- get_pseudobulk(meta, repository=NULL,cache_directory = "~/projects/caq/pseudobulk/0.2.1")
+#' }
+#'
+#' @importFrom dplyr pull filter as_tibble inner_join collect
+#' @importFrom tibble column_to_rownames
+#' @importFrom purrr reduce map map_int imap keep
+#' @importFrom BiocGenerics cbind
+#' @importFrom glue glue
+#' @importFrom SummarizedExperiment colData assayNames<-
+#' @importFrom httr parse_url
+#' @importFrom assertthat assert_that has_name
+#' @importFrom cli cli_alert_success cli_alert_info
+#' @importFrom rlang .data
+#' @importFrom S4Vectors DataFrame
+#' @export
+get_pseudobulk <- function(
+    data,
+    assays = "counts",
+    cache_directory = get_default_cache_dir(),
+    repository = NULL, # cloud container not ready yet
+    features = NULL
+) {
+  # Parameter validation
+  assays %in% names(assay_map) |>
+    all() |>
+    assert_that(
+      msg = 'assays must be a character vector containing "counts" and/or
+            "cpm"'
+    )
+  assert_that(
+    !anyDuplicated(assays),
+    inherits(cache_directory, "character"),
+    is.null(repository) || is.character(repository),
+    is.null(features) || is.character(features)
+  )
+  
+  # Data parameter validation (last, because it's slower)
+  ## Evaluate the promise now so that we get a sensible error message
+  force(data)
+  ## We have to convert to an in-memory table here, or some of the dplyr
+  ## operations will fail when passed a database connection
+  cli_alert_info("Realising metadata.")
+  raw_data <- collect(data)
+  assert_that(
+    inherits(raw_data, "tbl"),
+    has_name(raw_data, c("sample_"))
+  )
+  
+  versioned_cache_directory <- cache_directory
+  versioned_cache_directory |> dir.create(
+    showWarnings = FALSE,
+    recursive = TRUE
+  )
+  
+  subdirs <- assay_map[assays]
```

Comment:
There is a lot of duplication of validation logic and other things here. I would rather see this refactored into a separate function that is called by both `get_single_cell_experiment` and `get_pseudobulk`.